Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs

```diff
@@ -0,0 +1,280 @@
+            {
+                return Token;
+            }
+        }
+
+        Task<bool> ITokenAccess.RequestAccessTokenAsync(CancellationToken taskCancellationToken)
+        {
+            return RequestAccessTokenAsync(taskCancellationToken);
+        }
+
+        #endregion
+
+        /// <summary>Provides access to the underlying credential object</summary>
+        internal abstract object UnderlyingCredential { get; }
+
+        // We're explicitly implementing all the interfaces to only expose the members user actually
+        // needs to see. Because you cannot make explicit interface implementors abstract, they are redirecting
+        // to the following protected abstract members.
+        protected abstract void Initialize(ConfigurableHttpClient httpClient);
+
+        protected abstract TokenResponse Token { get; }
+
+        protected abstract Task<bool> RequestAccessTokenAsync(CancellationToken taskCancellationToken);
+
+        #region Factory methods
+
+        internal static GoogleCredential FromCredential(ComputeCredential credential)
+        {
+            return new ComputeGoogleCredential(credential);
+        }
+
+        internal static GoogleCredential FromCredential(ServiceAccountCredential credential)
+        {
+            return new ServiceAccountGoogleCredential(credential);
+        }
+
+        internal static GoogleCredential FromCredential(UserCredential credential)
+        {
+            return new UserGoogleCredential(credential);
+        }
+
+        #endregion
+
+        /// <summary>
+        /// Wraps <c>ComputeCredential</c> as <c>GoogleCredential</c>.
+        /// </summary>
+        internal class ComputeGoogleCredential : GoogleCredential
+        {
+            private readonly ComputeCredential credential;
+
+            public ComputeGoogleCredential(ComputeCredential credential)
+            {
+                this.credential = credential;
+            }
+
+            #region GoogleCredential overrides
+
+            internal override object UnderlyingCredential
+            {
+                get
+                {
+                    return credential;
+                }
+            }
+
+            protected override void Initialize(ConfigurableHttpClient httpClient)
+            {
+                credential.Initialize(httpClient);
+            }
+
+            protected override TokenResponse Token
+            {
+                get
+                {
+                    return credential.Token;
+                }
+            }
+
+            protected override Task<bool> RequestAccessTokenAsync(CancellationToken taskCancellationToken)
+            {
+                return RequestAccessTokenAsync(taskCancellationToken);
+            }
+
+            #endregion
+        }
+
+        /// <summary>
+        /// Wraps <c>ServiceAccountCredential</c> as <c>GoogleCredential</c>.
+        /// </summary>
+        internal class ServiceAccountGoogleCredential : GoogleCredential
```

Comment:
LGTM on the changes. This is close to the API surface that we pushed for originally.